In [138]:
import numpy as np

In [124]:
import pandas as pd
colnames = ['surgery?', 
'Age', 
'Hospital Number', 
'rectal temperature', 
'pulse', 'respiratory rate',
'temperature of extremities',
'peripheral pulse',
'mucous membranes',
'capillary refill time',
'pain',
'peristalsis',
'abdominal distension',
'asogastric tube',
'nasogastric reflux',
'nasogastric reflux PH',
'rectal examination - feces',
'abdomen',
'packed cell volume',
'total protein',
'abdominocentesis appearance',
'abdomcentesis total protein',
'outcome',
'surgical lesion?',
'type of lesion1',
'type of lesion2',
'type of lesion3',
'cp_data']

horse_data = pd.read_csv('horse_data.csv', names=colnames)
clean = horse_data.drop(columns = ['Hospital Number',
'peripheral pulse',
'mucous membranes',
'capillary refill time',
'peristalsis',
'abdominal distension',
'asogastric tube',
'nasogastric reflux',
'nasogastric reflux PH',
'rectal examination - feces',
'abdomen',
'packed cell volume',
'total protein',
'abdominocentesis appearance',
'abdomcentesis total protein',
'surgical lesion?',
'type of lesion1',
'type of lesion2',
'type of lesion3',
'cp_data'], axis = 1)
clean

surgery?  Age rectal temperature pulse respiratory rate  \
0          2    1              38.50    66               28   
1          1    1               39.2    88               20   
2          2    1              38.30    40               24   
3          1    9              39.10   164               84   
4          2    1              37.30   104               35   
..       ...  ...                ...   ...              ...   
295        1    1                  ?   120               70   
296        2    1              37.20    72               24   
297        1    1              37.50    72               30   
298        1    1              36.50   100               24   
299        1    1               37.2    40               20   

    temperature of extremities pain outcome  
0                            3    5       2  
1                            ?    3       3  
2                            1    3       1  
3                            4    2       2  
4                            ?    ?       2  
..                         ...  ...     ...  
295                          4    2       3  
296                          3    4       3  
297                          4    4       2  
298                          3    3       1  
299                          ?    ?       3  

[300 rows x 8 columns]

In [139]:
clean = clean.replace(to_replace='?', value=np.nan)
clean = clean.astype(float)
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    float64
 2   rectal temperature          281 non-null    float64
 3   pulse                       291 non-null    float64
 4   respiratory rate            242 non-null    float64
 5   temperature of extremities  300 non-null    float64
 6   pain                        300 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(8)
memory usage: 18.9 KB


In [140]:
clean.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          281.000000  291.000000   
mean     1.397993    1.640000           38.171708   72.288660   
std      0.490305    2.173972            0.685377   28.161343   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.100000   64.000000   
75%      2.000000    1.000000           38.500000   90.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        242.000000                  300.000000  300.000000  299.000000  
mean          30.417355                    2.296667    3.010000    1.551839  
std           17.642231                    1.041925    1.260368    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           18.500000                    1.000000    2.000000    1.000000  
50%           24.500000                    3.000000    3.000000    1.000000  
75%           36.000000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

Заполняем значения показателей ректальной температуры, пульса и частоты дыхания с учетом взаимосвязи со столбцами temperature of extremities и pain. Описание датасета говорит о том, что именно эти два столбца ранжируют лошадей на больных (по ощущению боли, шока), поэтому так будут корректно заполнятся ненормальные и нормальные характеристики лошадей.

In [141]:
clean['rectal temperature'] = clean['rectal temperature'].fillna(clean.groupby(['temperature of extremities','pain'])['rectal temperature'].transform('median'))

In [142]:
clean['respiratory rate'] = clean['respiratory rate'].fillna(clean.groupby(['temperature of extremities','pain'])['respiratory rate'].transform('median'))

In [143]:
clean['pulse'] = clean['pulse'].fillna(clean.groupby(['temperature of extremities','pain'])['pulse'].transform('median'))

In [144]:
clean

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0  1.0                38.5   66.0              28.0   
1         1.0  1.0                39.2   88.0              20.0   
2         2.0  1.0                38.3   40.0              24.0   
3         1.0  9.0                39.1  164.0              84.0   
4         2.0  1.0                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0  1.0                38.5  120.0              70.0   
296       2.0  1.0                37.2   72.0              24.0   
297       1.0  1.0                37.5   72.0              30.0   
298       1.0  1.0                36.5  100.0              24.0   
299       1.0  1.0                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           1.0   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           3.0   5.0      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         3.0   2.0      3.0  

[300 rows x 8 columns]

Заполняю столбец pain в зависимости от outcome, а столбец temperature of extremities - в зависимости от столбца pain по самому повторяющемуся значению. Исход зависит от уровня боли, а уровень температуры конечностей свидетельствует о состоянии лошдади (cool to cold extremities indicate possible shock)

In [145]:
clean['pain'] = clean['pain'].fillna(clean.groupby(['outcome'])['pain'].transform(lambda x:x.value_counts().index[0]))
clean

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0  1.0                38.5   66.0              28.0   
1         1.0  1.0                39.2   88.0              20.0   
2         2.0  1.0                38.3   40.0              24.0   
3         1.0  9.0                39.1  164.0              84.0   
4         2.0  1.0                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0  1.0                38.5  120.0              70.0   
296       2.0  1.0                37.2   72.0              24.0   
297       1.0  1.0                37.5   72.0              30.0   
298       1.0  1.0                36.5  100.0              24.0   
299       1.0  1.0                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           1.0   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           3.0   5.0      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         3.0   2.0      3.0  

[300 rows x 8 columns]

In [146]:
clean['temperature of extremities'] = clean['temperature of extremities'].fillna(clean.groupby(['pain'])['temperature of extremities'].transform(lambda x:x.value_counts().index[0]))
clean

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0         2.0  1.0                38.5   66.0              28.0   
1         1.0  1.0                39.2   88.0              20.0   
2         2.0  1.0                38.3   40.0              24.0   
3         1.0  9.0                39.1  164.0              84.0   
4         2.0  1.0                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0  1.0                38.5  120.0              70.0   
296       2.0  1.0                37.2   72.0              24.0   
297       1.0  1.0                37.5   72.0              30.0   
298       1.0  1.0                36.5  100.0              24.0   
299       1.0  1.0                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           1.0   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           3.0   5.0      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         3.0   2.0      3.0  

[300 rows x 8 columns]

In [147]:
clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    299 non-null    float64
 1   Age                         300 non-null    float64
 2   rectal temperature          299 non-null    float64
 3   pulse                       300 non-null    float64
 4   respiratory rate            300 non-null    float64
 5   temperature of extremities  300 non-null    float64
 6   pain                        300 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(8)
memory usage: 18.9 KB


Пропущенные значения в столбцах surgery и outcome считаю необходимым удалить, так мы не можем воспольнить их значения по соседним характеристикам - нельзя понять, прошла ли операция у лошади, так как другие характеристики могут показывать ее состояние, как до, так и после вмешательства. Также нельзя понять, умерла ли лошадь или еще нет. 

In [149]:
clean.dropna(inplace=True)
clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 298 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery?                    298 non-null    float64
 1   Age                         298 non-null    float64
 2   rectal temperature          298 non-null    float64
 3   pulse                       298 non-null    float64
 4   respiratory rate            298 non-null    float64
 5   temperature of extremities  298 non-null    float64
 6   pain                        298 non-null    float64
 7   outcome                     298 non-null    float64
dtypes: float64(8)
memory usage: 21.0 KB


1:  surgery?
          1 = Yes, it had surgery
          2 = It was treated without surgery

  2:  Age
          1 = Adult horse
          2 = Young (< 6 months)

  4:  rectal temperature
          - linear
          - in degrees celsius.
          - An elevated temp may occur due to infection.
          - temperature may be reduced when the animal is in late shock
          - normal temp is 37.8
          - this parameter will usually change as the problem progresses
               eg. may start out normal, then become elevated because of
                   the lesion, passing back through the normal range as the
                   horse goes into shock
  5:  pulse
          - linear
          - the heart rate in beats per minute
          - is a reflection of the heart condition: 30 -40 is normal for adults
          - rare to have a lower than normal rate although athletic horses
            may have a rate of 20-25
          - animals with painful lesions or suffering from circulatory shock
            may have an elevated heart rate

  6:  respiratory rate
          - linear
          - normal rate is 8 to 10
          - usefulness is doubtful due to the great fluctuations

  7:  temperature of extremities
          - a subjective indication of peripheral circulation
          - possible values:
               1 = Normal
               2 = Warm
               3 = Cool
               4 = Cold
          - cool to cold extremities indicate possible shock
          - hot extremities should correlate with an elevated rectal temp.
          
 11: pain - a subjective judgement of the horse's pain level
          - possible values:
               1 = alert, no pain
               2 = depressed
               3 = intermittent mild pain
               4 = intermittent severe pain
               5 = continuous severe pain
          - should NOT be treated as a ordered or discrete variable!
          - In general, the more painful, the more likely it is to require
            surgery
          - prior treatment of pain may mask the pain level to some extent

 23: outcome
          - what eventually happened to the horse?
          - possible values:
               1 = lived
               2 = died
               3 = was euthanized
